### ISSUE TITLE
BUG: can't grid search optimizers

### ISSUE THREAD

**Issue by adriangb** [adriangb/scikeras#70]
It turns out that Keras optimizers are not picklable because they use lambdas inside functions 🤦 

This can be fixed by using `tf.keras.optimizers.serialize` and `tf.keras.optimizers.deserialize`.

I think that maybe we need to list out all of the keras object types `Model`, `optimizers`, `callbacks`, etc. and use `copyreg.pickle` to register how to pickle each one using their custom keras serialize/deserialize methods.

**Comment by stsievert** [comment]
This issue might be worth reporting upstream, especially because pickle is a protocol. That means they could use `tf.keras.optimizers.{serialize, deserialize}` inside the pickle implementation (`__reduce__` or `__{set, get}state__`).

The post "[Pickle isn't slow, it's a protocol][1]" is somewhat related, and has a relevant example of using PyTorch serialization inside their Tensor class.

[1]:https://matthewrocklin.com/blog/work/2018/07/23/protocols-pickle

**Comment by adriangb** [comment]
I totally agree with you that the answer is for TF to implement the pickle protocol using it's own serialize/deserialzie functions (which I presume are optimized). I already suggested something similar to the TF team at https://github.com/tensorflow/tensorflow/pull/39609. I could open PRs for every single place this could be implemented, but I'm not sure that that will make things go any faster.

That post smells just like this problem. It's unfortunate that the TF maintainers are a lot less nimble and flexible that PyTorch.

**Comment by adriangb** [comment]
I left another comment in the TF PR. We'll see if we get a response. I'll give it a week or two and if not go ahead and implement the monkey patching here.

**Comment by adriangb** [comment]
So the RFC was approved on the Keras end, but it did not cover the specific use case here, which includes `optimizers`, `losses`, `metrics` and `callbacks`. The pushback on these was mainly surrounding optimizers, given that there is no singular "official" API for serializing optimizers _with_ their weights. Instead, we would have to call `tf.keras.optimizers.Optimizer.get_weights/tf.keras.optimizers.Optimizer.set_weights`.

We may not be able to upstream this since there is no clear use case in Keras for serializing optimizers, callbacks, etc. _outside_ of a `Model`.

**Comment by stsievert** [comment]
Thanks for commenting.

> given that there is no singular "official" API for serializing optimizers with their weights

That took me by surprise in the Google call. I think SciKeras should serialize optimizers. I think models that are serialized repeatedly should produce equivalent results to a model that is not serialized. That is, I'd like to see this test pass:

``` python
from sklearn.base import clone

m1 = KerasClassifier(..., optimizer="adam")
m2 = clone(m1)
for k in range(10):
    m1.partial_fit(X, y)
    tmp = pickle.loads(pickle.dumps(m2))
    assert tmp is not m2
    m2 = tmp
    m2.partial_fit(X, y)
    assert weights_close(m1, m2)

ypred1 = m1.predict(X)
ypred2 = m2.predict(X)
assert np.allclose(ypred1, ypred2)
```

I'm unsure if this test passes, or if it exists.

**Comment by adriangb** [comment]
Currently, the way we serialize models, I don't know if the optimizer weights are retained (that is, I don't know if `Model.get_weights` also gets the optimizer weights, I'm guessing not).

**Comment by stsievert** [comment]
Models should have weights. Optimizers should have gradients for the weights and some (small) internal state. I would suspect that the test in https://github.com/adriangb/scikeras/issues/70#issuecomment-723151716 passes if the gradients are re-initialized to zero after pickling (presuming the optimizer and model are correctly connected).

**Comment by adriangb** [comment]
That test should pass even if we did not implement anything special for pickling optimizers, as long as either (1) your assumption is correct or (2) SaveModel saves the entire state of optimizers.

The case where we might need to implement something specific for optimizers is:
```python3
from sklearn.base import clone

m1 = KerasClassifier(..., random_state=0, optimizer=optimizer_object)  # optimizer_object has _some_ state
m2 = pickle.loads(pickle.dumps(m1))  # state of optimizer_object may be lost

m1.fit(X, y)  # model_ is built with optimizer_object WITH original state
m2.fit(X, y)  # model_ is built with optimizer_object WITHOUT original state
assert weights_close(m1, m2)  # will fail
```

**Comment by adriangb** [comment]
Resolved by #128
